internal/charmstore/resources_test.go

```diff
@@ -175,6 +177,222 @@ func (s *resourceSuite) TestListResourcesResourceNotFound(c *gc.C) {
+		params.NoChannel,
+	} {
+		res, err := store.UploadResource(entity, "for-install", strings.NewReader("fake content"), fakeBlobHash, fakeBlobSize)
+		c.Assert(err, jc.ErrorIsNil)
+		resources = append(resources, res)
+		if ch == params.NoChannel {
+			continue
+		}
+		err = store.PublishResources(entity, ch, []mongodoc.ResourceRevision{{res.Name, res.Revision}})
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	for i, test := range resolveResourceTests {
+		c.Logf("%d. %s", i, test.about)
+		res, err := store.ResolveResource(EntityResolvedURL(entity), test.name, test.revision, test.channel)
+		if test.expectError != "" {
+			c.Assert(err, gc.ErrorMatches, test.expectError)
+			c.Assert(errgo.Cause(err), gc.Equals, params.ErrNotFound)
+			continue
+		}
+		c.Assert(err, jc.ErrorIsNil)
+		adjustExpectedResource(res, resources[test.expectResource])
+		c.Assert(res, jc.DeepEquals, resources[test.expectResource])
+	}
+}
+
+var publishResourceErrorTests = []struct {
+	about       string
+	channel     params.Channel
+	resources   []mongodoc.ResourceRevision
+	expectError string
+}{{
+	about:       "no channel",
+	channel:     params.NoChannel,
+	expectError: `missing channel`,
+}, {
+	about:       "unpublished channel",
+	channel:     params.UnpublishedChannel,
+	expectError: `cannot publish to unpublished channel`,
+}, {
+	about:   "unknown resource name",
+	channel: params.StableChannel,
+	resources: []mongodoc.ResourceRevision{{
+		Name:     "for-install",
+		Revision: 0,
+	}, {
+		Name:     "bad-name",
+		Revision: 0,
+	}},
+	expectError: `charm does not have resource "bad-name"`,
+}, {
+	about:   "no such resource",
+	channel: params.StableChannel,
+	resources: []mongodoc.ResourceRevision{{
+		Name:     "for-install",
+		Revision: 0,
+	}},
+	expectError: `cs:~charmers/xenial/starsay-3 resource for-install revison 0 not found`,
+}}
+
+func (s *resourceSuite) TestPublishResourceErrors(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, _ := addCharm(c, store, curl)
+	for i, test := range publishResourceErrorTests {
+		c.Logf("%d. %d", i, test.about)
+		err := store.PublishResources(entity, test.channel, test.resources)
+		c.Assert(err, gc.ErrorMatches, test.expectError)
+	}
+}
+
+func (s *resourceSuite) TestOpenResourceBlob(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, _ := addCharm(c, store, curl)
+	res, err := store.UploadResource(entity, "for-install", strings.NewReader("fake content"), fakeBlobHash, fakeBlobSize)
+	c.Assert(err, jc.ErrorIsNil)
+	blob, err := store.OpenResourceBlob(res)
+	c.Assert(err, jc.ErrorIsNil)
+	defer blob.Close()
+	c.Assert(blob.Size, gc.Equals, fakeBlobSize)
+	c.Assert(blob.Hash, gc.Equals, fakeBlobHash)
+	data, err := ioutil.ReadAll(blob)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(string(data), gc.Equals, "fake content")
+	res.BlobName = res.BlobName[1:]
```

Comment:
// Change the blob name so that it's invalid
// so that we can check what happens then.

?